### ISSUE TITLE
We should remove -Werror=shadow from the CFLAGS when we detect we run on an older gcc version

### ISSUE THREAD

**Issue by utezduyar** [systemd/systemd#406]
I am not sure why you guys don't see these errors but I see many shadow errors since we enabled the switch. I started fixing some but then decided to ask first. Is it our compiler acting crazy? 

```
shadow/src/libsystemd/sd-bus/bus-kernel.c: In function 'bus_message_setup_kmsg':
shadow/src/libsystemd/sd-bus/bus-kernel.c:323:33: error: declaration of 'now' shadows a global declaration [-Werror=shadow]
In file included from shadow/src/basic/util.h:47:0,
                 from shadow/src/libsystemd/sd-bus/bus-kernel.c:37:
shadow/src/basic/time-util.h:72:8: error: shadowed declaration is here [-Werror=shadow]


shadow/src/basic/hashmap.c: In function 'get_hash_key':
shadow/src/basic/hashmap.c:386:34: error: declaration of 'hash_key' shadows a global declaration [-Werror=shadow]
shadow/src/basic/hashmap.c:376:17: error: shadowed declaration is here [-Werror=shadow]

shadow/src/basic/rm-rf.c: In function 'rm_rf_children':
shadow/src/basic/rm-rf.c:65:22: error: declaration of 'is_dir' shadows a global declaration [-Werror=shadow]
In file included from shadow/src/basic/rm-rf.c:22:0:
shadow/src/basic/util.h:838:5: error: shadowed declaration is here [-Werror=shadow]

shadow/src/shared/fstab-util.h: In function 'fstab_test_yes_no_option':
shadow/src/shared/fstab-util.h:39:75: error: declaration of 'yes_no' shadows a global declaration [-Werror=shadow]
In file included from shadow/src/shared/generator.c:24:0:
shadow/src/basic/util.h:87:27: error: shadowed declaration is here [-Werror=shadow]

shadow/src/core/device.c: In function 'device_update_found_one':
shadow/src/core/device.c:444:82: error: declaration of 'now' shadows a global declaration [-Werror=shadow]
In file included from shadow/src/basic/util.h:47:0,
                 from shadow/src/basic/hashmap.h:28,
                 from shadow/src/shared/install.h:32,
                 from shadow/src/core/unit.h:36,
                 from shadow/src/core/dbus-device.h:24,
                 from shadow/src/core/device.c:28:
shadow/src/basic/time-util.h:72:8: error: shadowed declaration is here [-Werror=shadow]
shadow/src/core/device.c: In function 'device_update_found_by_sysfs':
shadow/src/core/device.c:476:106: error: declaration of 'now' shadows a global declaration [-Werror=shadow]
In file included from shadow/src/basic/util.h:47:0,
                 from shadow/src/basic/hashmap.h:28,
                 from shadow/src/shared/install.h:32,
                 from shadow/src/core/unit.h:36,
                 from shadow/src/core/dbus-device.h:24,
                 from shadow/src/core/device.c:28:
shadow/src/basic/time-util.h:72:8: error: shadowed declaration is here [-Werror=shadow]
shadow/src/core/device.c: In function 'device_update_found_by_name':
shadow/src/core/device.c:492:104: error: declaration of 'now' shadows a global declaration [-Werror=shadow]
In file included from shadow/src/basic/util.h:47:0,
                 from shadow/src/basic/hashmap.h:28,
                 from shadow/src/shared/install.h:32,
                 from shadow/src/core/unit.h:36,
                 from shadow/src/core/dbus-device.h:24,
                 from shadow/src/core/device.c:28:
shadow/src/basic/time-util.h:72:8: error: shadowed declaration is here [-Werror=shadow]
shadow/src/core/device.c: In function 'device_found_node':
shadow/src/core/device.c:767:87: error: declaration of 'now' shadows a global declaration [-Werror=shadow]
In file included from shadow/src/basic/util.h:47:0,
                 from shadow/src/basic/hashmap.h:28,
                 from shadow/src/shared/install.h:32,
                 from shadow/src/core/unit.h:36,
                 from shadow/src/core/dbus-device.h:24,
                 from shadow/src/core/device.c:28:
shadow/src/basic/time-util.h:72:8: error: shadowed declaration is here [-Werror=shadow]
```

**Comment by poettering** [comment]
On Mon, 29.06.15 06:41, Umut Tezduyar Lindskog (notifications@github.com) wrote:

> I am not sure why you guys don't see these errors but I see many
> shadow errors since we enabled the switch. I started fixing some but
> then decided to ask first. Is it our compiler acting crazy?

Some older gcc versions complain if you have a local variable and
a global function sharing the same name. Newer gcc versions don't do
that anymore. 

We should probably drop the -Werror=shadow automatically when we
notice that the local gcc versions acts crazy about it. I'd be happy
to merge a patch that adds a configure check for that!

Lennart

## 

Lennart Poettering, Red Hat

**Comment by poettering** [comment]
The configure check should probably just try to compile a program with -Werror-shadow that looks like this:

```
#include <unistd.h>

static int main(int argc, char *argv[]) {
        int unlink = 7;
}
```

With -Werror=shadow this would fail on older gcc, but work on newer gcc, using the fact that unlink() is a function defined by unistd.h.

**Comment by zabbal** [comment]
Which gcc version introduced this? Maybe it would be easier to just up the requirements for minimal gcc version?

**Comment by poettering** [comment]
I generally prefer checks for behaviour over checks for version numbers, since distros tend to backport features...

**Comment by utezduyar** [comment]
I am really horrible at autotools (who isn't). For this reason, until I find a time to implement ./configure check, I will have a downstream patch for this. I though, I shouldn't hold up anyone in case someone else wants to implement this.

**Comment by utezduyar** [comment]
I think similar thing with implicit-function-declaration.

Regardless of preparing a patch as Lennart said, if you want a quick fix, `cc_cv_CFLAGS__Werror_implicit_function_declaration=no` solves it too.

**Comment by dvdhrm** [comment]
This was fixed with gcc-4.8 (released March, 2013). This is >2 years old, so usually falls outside of the timeframe we support in -git. I think we'd still happily merge an autoconf-test upstream, but I doubt anyone of us will come up with something.

Following our usual rule of requiring deps that're <=2years old, I'll go ahead and close this. If anyone cares, please send a patch. But this is now pending for >1 month, so there's little point in keeping it opened.
